web/api/v1/api.go

```diff
@@ -477,13 +477,93 @@ func (api *API) labelNames(r *http.Request) apiFuncResult {
 	return apiFuncResult{names, nil, warnings, nil}
 }
 
+var (
+	minTime = time.Unix(math.MinInt64/1000+62135596801, 0).UTC()
+	maxTime = time.Unix(math.MaxInt64/1000-62135596801, 999999999).UTC()
+
+	minTimeFormatted = minTime.Format(time.RFC3339Nano)
+	maxTimeFormatted = maxTime.Format(time.RFC3339Nano)
+)
+
+func parseTimeParam(r *http.Request, paramName string, defaultValue time.Time) (time.Time, error) {
+	val := r.FormValue(paramName)
+	if val != "" {
+		result, err := parseTime(val)
+		if err != nil {
+			return result, errors.Wrapf(err, "Invalid parameter '%s'", paramName)
+		}
+		return result, nil
+	}
+	return defaultValue, nil
+}
+
+func selectSeriesByMatchers(q storage.Querier, matcherSets [][]*labels.Matcher) ([]storage.SeriesSet, error, storage.Warnings) {
+	var sets []storage.SeriesSet
+	var warnings storage.Warnings
+	for _, mset := range matcherSets {
+		s, wrn, err := q.Select(nil, mset...)
+		warnings = append(warnings, wrn...)
+		if err != nil {
+			return nil, err, warnings
+		}
+		sets = append(sets, s)
+	}
+
+	return sets, nil, nil
+}
+
 func (api *API) labelValues(r *http.Request) apiFuncResult {
+	if err := r.ParseForm(); err != nil {
+		return apiFuncResult{nil, &apiError{errorBadData, errors.Wrapf(err, "error parsing form values")}, nil, nil}
+	}
+
 	ctx := r.Context()
 	name := route.Param(ctx, "name")
 
 	if !model.LabelNameRE.MatchString(name) {
 		return apiFuncResult{nil, &apiError{errorBadData, errors.Errorf("invalid label name: %q", name)}, nil, nil}
 	}
+
+	if len(r.Form["match[]"]) > 0 {
+		start, err := parseTimeParam(r, "start", minTime)
+		if err != nil {
+			return apiFuncResult{nil, &apiError{errorBadData, err}, nil, nil}
+		}
+
+		end, err := parseTimeParam(r, "end", maxTime)
+		if err != nil {
+			return apiFuncResult{nil, &apiError{errorBadData, err}, nil, nil}
+		}
+
+		selector, err := promql.ParseMetricSelector(r.FormValue("match[]"))
+		if err != nil {
+			return apiFuncResult{nil, &apiError{errorBadData, err}, nil, nil}
+		}
+
+		q, err := api.Queryable.Querier(ctx, timestamp.FromTime(start), timestamp.FromTime(end))
+		defer q.Close()
+
+		if err != nil {
+			return apiFuncResult{nil, &apiError{errorExec, err}, nil, nil}
+		}
+
+		matches, err, warnings := selectSeriesByMatchers(q, [][]*labels.Matcher{selector})
+		if err != nil {
+			return apiFuncResult{nil, &apiError{errorExec, err}, warnings, nil}
+		}
+
+		series := storage.NewMergeSeriesSet(matches, nil)
+		result := make(map[string][]string)
+		for series.Next() {
+			label := series.At().Labels().Get(name)
+			if label != "" && result[label] == nil {
+				result[name] = append(result[name], label)
+				result[label] = result[name] // Prevent from adding same label value twice.
```

Comment:
I think we should be able to simplify this a bit, but lets have a look at the tests and will revisit